test/semant/Tests/ScopingSpec.hs

```diff
@@ -0,0 +1,263 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Tests.ScopingSpec where
+
+import           Sslang.Test
+
+import qualified Front
+import           Front.Scope                    ( scopeProgram )
+
+doScope :: String -> Pass ()
+doScope s = Front.parseAst def s >>= scopeProgram
+
+spec :: Spec
+spec = do
+  describe "scoping global definitions" $ do
+    it "supports global scoping" $ do
+      shouldPass $ doScope [here|
+        x = 3
+        y = x
+      |]
+    it "supports global function arguments" $ do
+      shouldPass $ doScope [here|
+        f x = x
+      |]
+    it "supports local and global scoping" $ do
+      shouldPass $ doScope [here|
+        foo x =
+          let foo' y = x
+          foo'
+        bar a b =
+          foo b a
+      |]
+    it "rejects undefined references" $ do
+      doScope [here|
+        f = x
+      |] `shouldFailWith` ScopeError "x is undefined"
```

Comment:
`"x is undefined"` here is an error message. Since error messages are always equal, we can write some descriptive error message about how we expect the test to fail, without worrying about the test being thwarted due to a failed string comparison.